### ISSUE TITLE
AtariWrapper does not use recommended defaults

### ISSUE THREAD

**Issue by RyanNavillus** [DLR-RM/stable-baselines3#635]
The current AtariWrapper by default has `terminate_on_life_loss` set to True. This goes against the recommendations of Revisiting the Arcade Learning Environment (https://arxiv.org/pdf/1709.06009.pdf). I believe this should be set to False by default. They also recommend using sticky actions instead of noop resets, but I think that problem is outside the scope of this wrapper.

**Comment by araffin** [comment]
> The current AtariWrapper by default has `terminate_on_life_loss` set to True. This goes against the recommendations of Revisiting the Arcade Learning Environment (https://arxiv.org/pdf/1709.06009.pdf). I believe this should be set to False by default. They also recommend using sticky actions instead of noop resets, but I think that problem is outside the scope of this wrapper.

I did quick experiments on Breakout with PPO, with and without terminal on life loss activated, and the current default seems to have a good impact on performance:
https://wandb.ai/openrlbenchmark/openrlbenchmark/reports/Atari-defaults-PPO---VmlldzoxOTA4NjUz

Results are not significant at all yet, only 3 runs, but it should be investigated further.

**Comment by qgallouedec** [comment]
What about sticky actions?

**Comment by araffin** [comment]
> What about sticky actions?

you mean its influence on performance?

I don't know, I think the main issue is that the changes were made without benchmark (in the paper, comparison is only partial) and it looks like it's still missing.
@RyanNavillus @JesseFarebro @pseudo-rnd-thoughts am I wrong?

**Comment by qgallouedec** [comment]
I meant, is it implemented? After digging in the code, I realize that sticky actions are enabled by default directly by ALE, but that for v0 and v5 (not for v4, which is used by sb3), see the [gym documentation](https://github.com/Farama-Foundation/Gymnasium/blob/7e885597633bd7d9730ddf7c9524fda5d960ecb9/docs/environments/atari.md#stochasticity) and the [ALE code](https://github.com/mgbellemare/Arcade-Learning-Environment/blob/49154a3e96f0858e5e9a1f41c89cfe18e6741468/src/python/gym.py#L149). It seems to me that sticky actions is used by most of the works, so shouldn't we provide a `StickyActionWrapper` for v4?

Somehow related: https://github.com/DLR-RM/rl-baselines3-zoo/issues/133#issuecomment-881647449

**Comment by pseudo-rnd-thoughts** [comment]
I believe the primary difference between `v0` and `v4` is if sticky actions is enabled. This is a confusing change that ale made 

https://gymnasium.farama.org/environments/atari/#version-history-and-naming-schemes

**Comment by RyanNavillus** [comment]
@araffin I think the goal of sticky actions was not to improve performance w.r.t reward but to produce a robust policy instead of a deterministic action sequence. They show that it training with sticky actions causes DQN to have roughly the same performance when you evaluate with sticky actions on and off, while methods designed to exploit determinism in the environment perform much worse when evaluated with stick actions on. They argue that sticky actions are better than other methods for various reasons in section 5.3.

TL;DR, sticky actions are the recommended way to prevent agents from abusing determinism, not a way to improve rewards.

**Comment by RyanNavillus** [comment]
There is a much weaker argument in the paper that we should not `terminate_on_life_loss` because that's environment-specific knowledge, so it algorithms evaluated with that setting will overfit more to Atari. They also argue that `terminate_on_life_loss` has a debatable effect on performance, but it seems like your experiments show that it can help.

**Comment by araffin** [comment]
> TL;DR, sticky actions are the recommended way to prevent agents from abusing determinism, not a way to improve rewards.

thanks for you answer =)
My question was not about checking that it improves performance, but rather just know the impact (positive or negative) on performance.
Because if you don't have a baseline (I think the paper mostly benchmarked DQN), then it's hard to compare things.
My point is more that the changes were made to Atari defaults, but we still have no idea what is the difference in performance and what are baselines results for those defaults. As shown by my quick experiments, there might be significant changes, but we currently don't know...

**Comment by RyanNavillus** [comment]
> we still have no idea what is the difference in performance and what are baselines results for those defaults

I might not be following exactly, why do we care about performance if the point of these changes is to prevent the agent from "cheating". I assume you would expect some drop in performance because both changes make it harder for the agent.

Is the issue that people are using existing SB3 results in their papers, and might mistakenly attribute the charts that you have now (without sticky actions) to the agents with stick actions if you change the defaults? I personally don't use results from a repository without rerunning the baselines myself, but I see why that could be a concern.

I think you're correct that the paper only benchmarks DQN/Sarsa on Atari games when evaluating sticky actions (In Table 8/9 they test their recommended settings on every Atari environment for those two algorithms). I'm sure there are other papers that use sticky actions for Atari baselines, but it would be a lot of work to find and verify those results. Probably faster to just run them yourself at that point. On that note, how much compute would it take to just run these baselines yourself?

**Comment by araffin** [comment]
> Is the issue that people are using existing SB3 results in their papers, and might mistakenly attribute the charts that you have now 

yes, that's the issue. And not only about SB3 results but all results published so far. You want to be able to compare apples to apples.

> I personally don't use results from a repository without rerunning the baselines myself, but I see why that could be a concern.

this is indeed good practice but [not everybody can afford doing so](http://proceedings.mlr.press/v139/ceron21a.html) (students/independent researchers/small labs/countries where compute is actually much more expensive/...).

> On that note, how much compute would it take to just run these baselines yourself?

if we want to run on all atari games, a lot... (you need 10 experiments x n_games x n_algorithms and this grows quickly...)
@qgallouedec is currently running a full benchmark of the RL Zoo, and we are using a subset of Atari games (https://github.com/openrlbenchmark/openrlbenchmark/issues/7), all runs is logged using W&B and help is welcomed =)

**Comment by RyanNavillus** [comment]
> this is indeed good practice but [not everybody can afford doing so](http://proceedings.mlr.press/v139/ceron21a.html) (students/independent researchers/small labs/countries where compute is actually much more expensive/...).

Sorry, I didn't intend for that to sound like a moral argument, I meant to say that I haven't worked on any papers that required it. I agree that having reliable baselines available is important for those reasons.

> if we want to run on all atari games, a lot... (you need 10 experiments x n_games x n_algorithms and this grows quickly...)
That is certainly a lot. It's unfortunate but without baselines like this, it's hard for the community to adopt the recommendations from the paper. Would it be possible to set up a system for people to contribute individual runs? I certainly can't afford all that compute myself, but I'd be willing to contribute runs where I can. Aside from sticky actions, I feel like that would be helpful for a bunch of things. If it's possible to export runs on wandb it shouldn't be too hard to do.

Aside from that, would it make sense to change the defaults and put a disclaimer on the existing Atari results that they were run with sticky actions and terminate on life loss disabled? That way new results will use the good defaults, but if people want to compare to old results they'll see the disclaimer and change their settings accordingly. Not perfect but it might be a step in the right direction.

**Comment by araffin** [comment]
> Would it be possible to set up a system for people to contribute individual runs? 

yes =D, that's the whole point of the openrl benchmark initiative by @vwxyzjn (best is probably to open an issue there to keep track of things and ask for permission to add runs to the sb3 namespace).

> Aside from that, would it make sense to change the defaults and put a disclaimer on the existing Atari results that they were run with sticky actions and terminate on life loss disabled? That way new results will use the good defaults,

my whole point for not changing is that I want the result first and change the default once we got the baselines (I also wouldn't call them "good default" but "new defaults").
But we can later tag all atari runs ran with v4 (see discussion in https://github.com/DLR-RM/rl-baselines3-zoo/pull/336).

**Comment by RyanNavillus** [comment]
> yes =D, that's the whole point of the openrl benchmark initiative by @vwxyzjn (best is probably to open an issue there to keep track of things and ask for permission to add runs to the sb3 namespace).

Oh awesome, I didn't realize anyone could contribute to that. I'll look into contributing my limited compute where I can :)

> I also wouldn't call them "good default" but "new defaults")

I think it's safe to call them good from the perspective of evaluation fairness and policy robustness, at least according to the (only?) paper discussing the impact of different ALE settings. Probably not for performance.

I'm not sure I agree with waiting to change the defaults, I think a disclaimer on the current results would be enough, but that's just my 2 cents. You know how this would impact SB3 users better than me.

**Comment by qgallouedec** [comment]
> I want the result first and change the default once we got the baselines

Agree, it's a good idea to have both in the benchmark (even if it duplicates the number of runs to do).

> I'm not sure I agree with waiting to change the defaults, I think a disclaimer on the current results would be enough

Also agree, I would recommend adding as soon as possible, a "sticky action" param, set to False at first so as not to change the default setting. But I agree with @RyanNavillus, in the shortest possible term, I think we should change default to sticky action = True, probably with a warning in the doc and in the code.

**Comment by vwxyzjn** [comment]
FWIW, here is the default setting for `openai/baselines` and the recommended setting:

```
envs = envpool.make(
    "Breakout-v5",
    env_type="gym",
    num_envs=8,
    episodic_life=True,
    repeat_action_probability=0.0,
    noop_max=30,
    full_action_space=False,
    # https://github.com/openai/gym/blob/a7b6462136ebaa610c8941e4da8a9c92155b04d1/gym/envs/__init__.py#L799
    # max_frames = 100000 * 4 = 400000
    max_episode_steps=100000,
    reward_clip=True,
    img_height=84,
    img_width=84,
    stack_num=4,
    gray_scale=True,
    frame_skip=4,
)
```
```
envs = envpool.make(
    "Breakout-v5",
    env_type="gym",
    num_envs=8,
    episodic_life=False,  # Machado et al. 2017 (Revisitng ALE: Eval protocols) p. 6
    repeat_action_probability=0.25,  # Machado et al. 2017 (Revisitng ALE: Eval protocols) p. 12
    noop_max=1,  # Machado et al. 2017 (Revisitng ALE: Eval protocols) p. 12 (no-op is deprecated in favor of sticky action, right?)
    full_action_space=True,  # Machado et al. 2017 (Revisitng ALE: Eval protocols) Tab. 5
    max_episode_steps=int(108000 / 4),  # Hessel et al. 2018 (Rainbow DQN), Table 3, Max frames per episode
    reward_clip=True,
    img_height=84,
    img_width=84,
    stack_num=4,
    gray_scale=True,
    frame_skip=4,
)
```

Maybe need to double-check if `noop_max` should be set to 0, but when I do, it gives me an error...  @Trinkle23897

**Comment by qgallouedec** [comment]
> (no-op is deprecated in favor of sticky action, right?)

No. In fact, some papers use no-ops ([NGU](https://arxiv.org/pdf/2002.06038.pdf), [Agent57](https://arxiv.org/pdf/2002.06038.pdf), [QR-DQN](https://arxiv.org/pdf/1710.10044v1.pdf), [R2D2](https://openreview.net/pdf?id=r1lyTjAqYX)), some use sticky actions ([Dreamerv2](https://arxiv.org/pdf/2010.02193v4.pdf)), and some use both, either combinating them ([Go-Explore](https://adrien.ecoffet.com/files/go-explore-nature.pdf)), or reporting result for both settings separated ([MuZero](https://arxiv.org/pdf/2104.06294v1.pdf)) . 


IMO, the best would be to use both as default. Justification from [Ecoffet et al. 2021](https://adrien.ecoffet.com/files/go-explore-nature.pdf) (Go-Explore): 

> As it is possible that no-ops add some challenges not encountered with just sticky actions, we ensure that Go-Explore is evaluated under conditions that are at least as difficult as those presented in recent papers by evaluating Go-Explore with both sticky actions and no-ops.



@araffin According to [Schrittwieser et al. 2021](https://arxiv.org/pdf/2104.06294v1.pdf) (MuZero), sticky action have low impact on score.

**Comment by qgallouedec** [comment]
> > reporting result for both settings separated ([MuZero](https://arxiv.org/pdf/2104.06294v1.pdf)) .
> 
> Could you point me to the page or table number for both settings?

Table 12 and section 7. After reading a second time, it is not clear if they use non-stop start in "MuZero sticky actions". Anyway, it doesn't matter.
